{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RiboMIMO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook contains the code for analyzing the results of RiboMIMO.\n",
    "\n",
    "Before analyzing, users need to:\n",
    "1. Train RiboMIMO model using one of the provided datasets, or a custom one with the same format;\n",
    "2. Calculate CIS using the trained model;\n",
    "3. Find the paths containing the train models and the computed CISs above.\n",
    "\n",
    "Please refer to Readme for detailed description.\n",
    "\n",
    "**The original results provided in this notebook is just demo for visualization using a small portion of results.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import json\n",
    "import argparse\n",
    "from argparse import Namespace\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "from scipy.interpolate import interpn\n",
    "from scipy.stats import spearmanr\n",
    "from runRiboMIMO import *\n",
    "os.environ[\"CUDA_DEVICE_ORDER\"] = \"PCI_BUS_ID\"  # see issue #152"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Loading results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Please specify the following items used for analysing\n",
    "\"\"\"\n",
    "# directory with trained models, [MODEL FOLDER]\n",
    "model_path = \"../results/RiboMIMO_Mohammad19-1_codon_nt_aa_0,2_3_256_2_sim_-1_bs_16_cv_1x10_train_20201229_115838/\"\n",
    "# directory with cis, [CIS FOLDER]\n",
    "cis_path = \"../results/CIS_Mohammad19-1_codon_nt_aa_0,2_3_256_2_sim_-1_bs_16_cv_1x10_cis_20201229_131846/\"\n",
    "# which gpu to use, \"-1\" for cpu\n",
    "gpu = \"0\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load config file\n",
    "config = json.load(open(model_path + \"config\", \"r\"))\n",
    "args = Namespace(**config)\n",
    "args.info = \"analysis\"\n",
    "args.gpu = gpu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reproduce results \n",
    "RiboMIMONet = RiboMIMOModel(args)\n",
    "logging.info(\"Dataset: {}\".format(RiboMIMONet.dataset))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Performance evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "list_length = []\n",
    "list_density = []\n",
    "list_pred = []\n",
    "list_fold = []\n",
    "np.random.seed(RiboMIMONet.seed)\n",
    "seeds = np.random.randint(0, 1024, RiboMIMONet.num_repeat)\n",
    "logging.info(\"{} repeat {} fold cross-validation\".format(RiboMIMONet.num_repeat, RiboMIMONet.num_fold))\n",
    "for repeat in range(RiboMIMONet.num_repeat):\n",
    "    RiboMIMONet.split_data(seeds[repeat])\n",
    "    for fold in range(RiboMIMONet.num_fold):\n",
    "        logging.info(\"{}/{} repeat {}/{} fold\".format(repeat+1, RiboMIMONet.num_repeat, fold+1, RiboMIMONet.num_fold))\n",
    "        RiboMIMONet.load_model()\n",
    "        RiboMIMONet.model.load_state_dict(torch.load(model_path+\"best_model_{}_{}.pth\".format(repeat, fold)))\n",
    "        RiboMIMONet.index_test = RiboMIMONet.data.list_fold[fold]\n",
    "        index_eval = RiboMIMONet.index_test\n",
    "        RiboMIMONet.index_train = RiboMIMONet.data.list_fold[:fold] + RiboMIMONet.data.list_fold[fold+1:]\n",
    "        RiboMIMONet.index_train = np.concatenate(RiboMIMONet.index_train)\n",
    "        RiboMIMONet.get_thresh()\n",
    "        RiboMIMONet.data.get_data_pack(index_eval, RiboMIMONet.add_nt, RiboMIMONet.add_aa)\n",
    "        RiboMIMONet.data.get_data_pack_cls(index_eval, RiboMIMONet.threshs)\n",
    "        num_batch = int(np.ceil(len(index_eval) / float(RiboMIMONet.batch_size)))\n",
    "        temp_density = [[] for _ in range(len(index_eval))]\n",
    "        temp_pred = [[] for _ in range(len(index_eval))]\n",
    "        with torch.no_grad():\n",
    "            for batch in range(num_batch):\n",
    "                index = np.array(range(batch*RiboMIMONet.batch_size, min((batch+1)*RiboMIMONet.batch_size,len(index_eval))))\n",
    "                x_input, length, density, mask_density, level, batch_index = RiboMIMONet.get_data_batch(index)\n",
    "                pred = RiboMIMONet.model.predict(x_input, length)\n",
    "                length = length.cpu().data.long()\n",
    "                for i in range(pred.shape[0]):\n",
    "                    temp_density[batch_index[i]] = density[i][0:length[i]].cpu().data\n",
    "                    temp_pred[batch_index[i]] = pred[i][0:length[i]].cpu().data\n",
    "        list_density.extend(temp_density)\n",
    "        list_pred.extend(temp_pred)\n",
    "        list_fold.extend([repeat*fold+fold] * len(temp_density))\n",
    "torch.cuda.empty_cache()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "list_results = []\n",
    "for idx, gene in enumerate(RiboMIMONet.data.list_gene):\n",
    "    list_results.append(RiboMIMONet.model.score_reg(list_pred[idx], list_density[idx]))\n",
    "list_results = np.array(list_results)\n",
    "printstring = \"\\033[91mr2: {:.6f}±{:.6f}, mse: {:.6f}±{:.6f}, corr: {:.6f}±{:.6f}\\033[0m\"\n",
    "logging.info(\"Per-gene correlation, SD across {} genes\".format(len(list_results)))\n",
    "logging.info(printstring.format(np.nanmean(list_results[:,0]), \n",
    "                                np.nanstd(list_results[:,0]),\n",
    "                                np.nanmean(list_results[:,1]), \n",
    "                                np.nanstd(list_results[:,1]),\n",
    "                                np.nanmean(list_results[:,2]), \n",
    "                                np.nanstd(list_results[:,2]),))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_results_cv = []\n",
    "list_fold = np.array(list_fold)\n",
    "for i in np.unique(list_fold):\n",
    "    list_results_fold = list_results[i == list_fold]\n",
    "    list_results_cv.append(np.nanmean(list_results_fold, 0))\n",
    "list_results_cv = np.array(list_results_cv)\n",
    "printstring = \"\\033[91mr2: {:.6f}±{:.6f}, mse: {:.6f}±{:.6f}, corr: {:.6f}±{:.6f}\\033[0m\"\n",
    "logging.info(\"Per-fold correlation, SD across {} folds\".format(len(list_results_cv)))\n",
    "logging.info(printstring.format(np.nanmean(list_results_cv[:,0]), \n",
    "                                np.nanstd(list_results_cv[:,0]),\n",
    "                                np.nanmean(list_results_cv[:,1]), \n",
    "                                np.nanstd(list_results_cv[:,1]),\n",
    "                                np.nanmean(list_results_cv[:,2]), \n",
    "                                np.nanstd(list_results_cv[:,2]),))\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. CIS analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1 Position to the ribosomal A sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "num = 0\n",
    "for seq in RiboMIMONet.data.dict_seq_codon.values():\n",
    "    if len(seq) > num:\n",
    "        num = int(len(seq))\n",
    "\n",
    "# 10 fold all saliency map\n",
    "list_distance_cis = np.zeros(num*2)\n",
    "counts = np.zeros(num*2)\n",
    "for idx in range(len(RiboMIMONet.data.list_gene)):\n",
    "    print(\"{:.2f}%\\r\".format((1.0+idx)/len(RiboMIMONet.data.list_gene)*100), end=\"\")\n",
    "    gene_name = RiboMIMONet.data.list_gene[idx]\n",
    "    seq = RiboMIMONet.data.dict_seq[gene_name]\n",
    "    len_codon = len(seq)//3\n",
    "    codons = [seq[i*3:(i+1)*3] for i in range(len_codon)]\n",
    "    try:\n",
    "        saliency_sum = pickle.load(open(\"{}/{}\".format(cis_path, gene_name), \"rb\"))\n",
    "    except:\n",
    "        continue\n",
    "    sm_2d = np.abs(saliency_sum)\n",
    "    for i in range(5, len_codon-4):\n",
    "        list_distance_cis[num+i-len_codon+6:num+i-5+1] += sm_2d[5:len_codon-5, i][::-1]\n",
    "        counts[num+i-len_codon+6:num+i-5+1] += 1\n",
    "list_distance_cis /= counts\n",
    "\n",
    "# 10 fold all saliency map\n",
    "list_distance_cis_sq1 = np.zeros(num*2)\n",
    "list_distance_cis_sq2 = np.zeros(num*2)\n",
    "counts1 = np.zeros(num*2)\n",
    "counts2 = np.zeros(num*2)\n",
    "for idx in range(len(RiboMIMONet.data.list_gene)):\n",
    "    print(\"{:.2f}%\\r\".format((1.0+idx)/len(RiboMIMONet.data.list_gene)*100), end=\"\")\n",
    "    gene_name = RiboMIMONet.data.list_gene[idx]\n",
    "    seq = RiboMIMONet.data.dict_seq[gene_name]\n",
    "    len_codon = len(seq)//3\n",
    "    codons = [seq[i*3:(i+1)*3] for i in range(len_codon)]\n",
    "    try:\n",
    "        saliency_sum = pickle.load(open(\"{}/{}\".format(cis_path, gene_name), \"rb\"))\n",
    "    except:\n",
    "        continue\n",
    "    sm_2d = np.abs(saliency_sum)\n",
    "    for i in range(5, len_codon-4):\n",
    "        control = list_distance_cis[num+i-len_codon+6:num+i-5+1]\n",
    "        error = sm_2d[5:len_codon-5, i][::-1]-control\n",
    "        list_distance_cis_sq1[num+i-len_codon+6:num+i-5+1] += (error**2) * (error > 0)\n",
    "        list_distance_cis_sq2[num+i-len_codon+6:num+i-5+1] += (error**2) * (error < 0)\n",
    "        counts1[num+i-len_codon+6:num+i-5+1] += error > 0\n",
    "        counts2[num+i-len_codon+6:num+i-5+1] += error < 0\n",
    "counts1[counts1==0] = 1\n",
    "counts2[counts2==0] = 1\n",
    "list_distance_cis_sq1 /= counts1\n",
    "list_distance_cis_sq2 /= counts2\n",
    "list_distance_cis_sq1[np.isnan(list_distance_cis_sq1)] = 0\n",
    "list_distance_cis_sq2[np.isnan(list_distance_cis_sq2)] = 0\n",
    "list_distance_cis_sq1 = np.sqrt(list_distance_cis_sq1)\n",
    "list_distance_cis_sq2 = np.sqrt(list_distance_cis_sq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 6), dpi=200)\n",
    "window = 14\n",
    "plotnum = 1000\n",
    "\n",
    "ax1.plot(list(range(-plotnum, plotnum)), \n",
    "         list_distance_cis[num-plotnum:num+plotnum], \n",
    "         color=\"C4\", alpha=1, lw=2)\n",
    "ax1.fill_between(list(range(-plotnum, plotnum)), \n",
    "                 list_distance_cis_sq1[num-plotnum:num+plotnum], \n",
    "                 list_distance_cis_sq2[num-plotnum:num+plotnum], \n",
    "                 alpha=0.5, color=\"C7\")\n",
    "ax1.legend([\"Mean\", \"SD\"], fontsize=20)\n",
    "\n",
    "ax1.set_xlabel(\"Offset from A site (codon)\", fontsize=20)\n",
    "ax1.set_ylabel(\"CIS$_d$\", fontsize=20)\n",
    "ax1.set_title(\"Contribution of positions near A site\\nin {}\".format(RiboMIMONet.dataset), \n",
    "              fontsize=20, pad=20)\n",
    "ax1.spines[\"left\"].set_position(('outward', 10))\n",
    "ax1.spines[\"bottom\"].set_position(('outward', 10))\n",
    "ax1.spines[\"top\"].set_visible(False)\n",
    "ax1.spines[\"right\"].set_visible(False)\n",
    "ax1.spines[\"bottom\"].set_bounds(-plotnum, plotnum)\n",
    "ax1.spines[\"left\"].set_bounds(0, 0.4)\n",
    "for tick in ax1.yaxis.get_major_ticks():\n",
    "    tick.label.set_fontsize(14) \n",
    "for tick in ax1.xaxis.get_major_ticks():\n",
    "    tick.label.set_fontsize(14) \n",
    "ax1.axis([-plotnum, plotnum, -0.02, 0.4])\n",
    "axis = ax1.axis()\n",
    "ax1.text(axis[0]-0.2*(axis[1]-axis[0]), \n",
    "         axis[3]+0.2*(axis[3]-axis[2]), \n",
    "         \"A\", fontsize=20, horizontalalignment='center', verticalalignment='center')\n",
    "ax1.yaxis.set_label_coords(-0.2, 0.5)\n",
    "\n",
    "\n",
    "ax2.bar(list(range(-window//2, window//2+1)), \n",
    "        list_distance_cis[num-window//2:num+window//2+1], \n",
    "        color=\"C4\", alpha=0.5)\n",
    "ax2.fill([-5.5, -5.5, 4.5, 4.5], [0, 0.3, 0.3, 0], color=\"C0\", alpha=0.1)\n",
    "ax2.set_xticks(range(-window//2, window//2+1))\n",
    "ax2.set_xlabel(\"Offset from A site (codon)\", fontsize=20)\n",
    "ax2.set_ylabel(\"CIS$_d$\", fontsize=20)\n",
    "ax2.set_title(\"Contribution of positions near A site\\nin {}\".format(RiboMIMONet.dataset), fontsize=20, pad=20)\n",
    "list_yticks = np.array([str(i) for i in range(-window//2, window//2+1)])\n",
    "list_yticks[window//2-2:window//2+1] = [\"E\", \"P\", \"A\"]\n",
    "ax2.set_xticklabels(list_yticks)\n",
    "ax2.spines[\"left\"].set_position(('outward', 10))\n",
    "ax2.spines[\"bottom\"].set_position(('outward', 10))\n",
    "for tick in ax2.xaxis.get_major_ticks():\n",
    "    tick.label.set_fontsize(14) \n",
    "for tick in ax2.yaxis.get_major_ticks():\n",
    "    tick.label.set_fontsize(14) \n",
    "ax2.spines[\"top\"].set_visible(False)\n",
    "ax2.spines[\"bottom\"].set_visible(False)\n",
    "ax2.spines[\"right\"].set_visible(False)\n",
    "ax2.spines[\"bottom\"].set_bounds(-window//2, window//2+1)\n",
    "ax2.spines[\"left\"].set_bounds(0, 0.4)\n",
    "ax2.tick_params(axis=\"x\", labelcolor=\"k\", colors=\"w\")\n",
    "ax2.axis([-window//2-1, window//2+1+1, -0.01, 0.4])\n",
    "axis = ax2.axis()\n",
    "ax2.text(axis[0]-0.2*(axis[1]-axis[0]), \n",
    "         axis[3]+0.2*(axis[3]-axis[2]), \n",
    "         \"B\", fontsize=20, horizontalalignment='center', verticalalignment='center')\n",
    "ax2.yaxis.set_label_coords(-0.2, 0.5)\n",
    "ax2.legend([\"Ribosome\\nprotected\\nregion\"], fontsize=20, loc=[0.55, 0.75])\n",
    "plt.subplots_adjust(left=0.125, bottom=0.1, right=0.9, top=0.9, wspace=0.3, hspace=0.2)\n",
    "# plt.show()\n",
    "plt.savefig(\"{}/Fig2_distance_{}.pdf\".format(RiboMIMONet.save_path, RiboMIMONet.dataset), bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Codon level contribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_codon = [nt1+nt2+nt3 for nt1 in \"TCAG\" for nt2 in \"TCAG\" for nt3 in \"TCAG\"]\n",
    "for codon in [\"TAA\", \"TAG\", \"TGA\"]:\n",
    "    list_codon.remove(codon)\n",
    "list_codon = np.array(list_codon)\n",
    "codon2idx = {list_codon[i]:i for i in range(len(list_codon))}\n",
    "list_aa = np.array([RiboMIMONet.data.codon2aa[codon] for codon in list_codon])\n",
    "index = np.argsort(list_aa)\n",
    "list_codon = list_codon[index]\n",
    "codon2idx = {list_codon[i]:i for i in range(len(list_codon))}\n",
    "list_aa = np.array([RiboMIMONet.data.codon2aa[codon] for codon in list_codon])\n",
    "list_aa_unique = np.unique(list_aa)\n",
    "aa2idx = {list_aa_unique[i]:i for i in range(len(list_aa_unique))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# CAI\n",
    "import pandas as pd\n",
    "codon_count = {}\n",
    "for idx in range(len(RiboMIMONet.data.list_gene)):\n",
    "    gene_name = RiboMIMONet.data.list_gene[idx]\n",
    "    seq = RiboMIMONet.data.dict_seq[gene_name]\n",
    "    len_codon = len(seq)//3\n",
    "    codon = [seq[i*3:(i+1)*3] for i in range(len_codon)]\n",
    "    for jdx in range(len_codon):\n",
    "        if codon[jdx] not in codon_count:\n",
    "            codon_count[codon[jdx]] = 1\n",
    "        else:\n",
    "            codon_count[codon[jdx]] += 1\n",
    "codon_usage = pd.DataFrame(columns=[\"Codon\", \"Number\", \"AA\"])\n",
    "codon_usage[\"Codon\"] = list(codon_count.keys())\n",
    "codon_usage[\"AA\"] = [RiboMIMONet.data.codon2aa[codon] for codon in list(codon_count.keys())]\n",
    "for i in codon_usage.index:\n",
    "    codon_usage.loc[i, \"Number\"] = codon_count[codon_usage.loc[i, \"Codon\"]]\n",
    "codon_usage_dict = {}\n",
    "for i in codon_usage.index:\n",
    "    aa = codon_usage.loc[i, \"AA\"]\n",
    "    fraction_max = codon_usage[\"Number\"][codon_usage[\"AA\"]== aa].max()\n",
    "    codon_usage_dict[codon_usage.loc[i, \"Codon\"]] = codon_usage.loc[i, \"Number\"]/fraction_max\n",
    "list_CAI = np.array([codon_usage_dict[codon] for codon in list_codon])\n",
    "used_info = \"CAI\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CIS\n",
    "site = \"A\"\n",
    "if site == \"A\":\n",
    "    site_offset = 0\n",
    "elif site == \"E\":\n",
    "    site_offset = 2\n",
    "dict_codon_cis = {}\n",
    "for idx in range(len(RiboMIMONet.data.list_gene)):\n",
    "    print(\"{:.2f}%\\r\".format((1.0+idx)/len(RiboMIMONet.data.list_gene)*100), end=\"\")\n",
    "    gene_name = RiboMIMONet.data.list_gene[idx]\n",
    "    seq = RiboMIMONet.data.dict_seq[gene_name]\n",
    "    len_codon = len(seq)//3\n",
    "    codons = [seq[i*3:(i+1)*3] for i in range(len_codon)]\n",
    "    try:\n",
    "        saliency_sum = pickle.load(open(\"{}/{}\".format(cis_path, gene_name), \"rb\"))\n",
    "    except:\n",
    "        continue\n",
    "    sm_2d = np.abs(saliency_sum)\n",
    "    for i in range(saliency_sum.shape[1]-site_offset):\n",
    "        if codons[i] not in dict_codon_cis:\n",
    "            dict_codon_cis[codons[i]] = [0, 0]\n",
    "        dict_codon_cis[codons[i]][0] += sm_2d[:,i+site_offset].sum()\n",
    "        dict_codon_cis[codons[i]][1] += sm_2d.shape[0]\n",
    "for codon in dict_codon_cis:\n",
    "    dict_codon_cis[codon] = dict_codon_cis[codon][0] / dict_codon_cis[codon][1]\n",
    "list_CIS = np.array([dict_codon_cis[codon] for codon in list_codon])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# density\n",
    "dict_density = {}\n",
    "for idx in range(len(RiboMIMONet.data.list_gene)):\n",
    "    gene_name = RiboMIMONet.data.list_gene[idx]\n",
    "    seq = RiboMIMONet.data.dict_seq[gene_name]\n",
    "    len_codon = len(seq)//3\n",
    "    codon = [seq[i*3:(i+1)*3] for i in range(len_codon)]\n",
    "    density = list_pred[idx].numpy()\n",
    "    assert len(density) == len(codon)\n",
    "    for jdx in range(len_codon-site_offset):\n",
    "        if codon[jdx+site_offset] not in dict_density:\n",
    "            dict_density[codon[jdx+site_offset]] = [np.abs(density[jdx]), np.array(density[jdx]>0, dtype=int)]\n",
    "        else:\n",
    "            dict_density[codon[jdx+site_offset]][0] += np.abs(density[jdx])\n",
    "            dict_density[codon[jdx+site_offset]][1] += np.array(density[jdx]>0, dtype=int)\n",
    "for codon in dict_density:\n",
    "    dict_density[codon] = dict_density[codon][0]/dict_density[codon][1]\n",
    "list_den = np.array([dict_density[codon] for codon in list_codon])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "focus = []\n",
    "for i in range(len(list_codon)):\n",
    "    if list_CIS[i]>0.01:\n",
    "        focus.append([list_codon[i], list_CIS[i], list_cAI[i], list_den[i]])\n",
    "focus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# saliency map, cAI\n",
    "used_info = \"CAI\"\n",
    "focus = []\n",
    "for i in range(len(list_codon)):\n",
    "    if list_CIS[i]>0.01:\n",
    "        focus.append([list_codon[i], list_CIS[i], list_CAI[i], list_den[i]])\n",
    "\n",
    "fig, (ax2, ax1) = plt.subplots(1, 2, figsize=(15, 6), dpi=200)\n",
    "scorr = spearmanr(list_den, list_CAI)\n",
    "pcorr = pearsonr(list_den, list_CAI)\n",
    "print(pcorr)\n",
    "print(scorr)\n",
    "ax2.scatter(list_CAI, list_den, color=\"C4\", marker=\".\", s=200)\n",
    "ax2.set_xlabel(used_info, fontsize=20)\n",
    "ax2.set_ylabel(\"Predicted ribosome density\", fontsize=20)\n",
    "ax2.set_title(\"Predicted ribosome density vs. {}\\nin {}\".format(used_info, RiboMIMONet.dataset), fontsize=20, pad=20)\n",
    "ax2.spines[\"top\"].set_visible(False)\n",
    "ax2.spines[\"right\"].set_visible(False)\n",
    "ax2.spines[\"left\"].set_position(('outward', 19))\n",
    "ax2.spines[\"bottom\"].set_position(('outward', 10))\n",
    "ax2.spines[\"bottom\"].set_bounds(0, 1)\n",
    "ax2.spines[\"left\"].set_bounds(0, 3.5)\n",
    "ax2.axis([-0.1, 1.1, 0, 3.5])\n",
    "for tick in ax2.xaxis.get_major_ticks():\n",
    "    tick.label.set_fontsize(14) \n",
    "for tick in ax2.yaxis.get_major_ticks():\n",
    "    tick.label.set_fontsize(14) \n",
    "axis = ax2.axis()\n",
    "ax2.text(axis[0]+0.3*(axis[1]-axis[0]), \n",
    "         axis[3]-0.05*(axis[3]-axis[2]), \n",
    "         \"scc: ${:.4f}$ (p-value: ${:.1g}$)\".format(scorr[0], scorr[1]), \n",
    "         fontsize=14, horizontalalignment='left', verticalalignment='center')\n",
    "ax2.text(axis[0]+0.3*(axis[1]-axis[0]), \n",
    "         axis[3]-0.1*(axis[3]-axis[2]), \n",
    "         \"pcc: ${:.4f}$ (p-value: ${:.1g}$)\".format(pcorr[0], pcorr[1]), \n",
    "         fontsize=14, horizontalalignment='left', verticalalignment='center')\n",
    "ax2.text(axis[0]-0.2*(axis[1]-axis[0]), \n",
    "         axis[3]+0.2*(axis[3]-axis[2]), \n",
    "         \"C\", fontsize=20, horizontalalignment='center', verticalalignment='center')\n",
    "ax2.yaxis.set_label_coords(-0.2, 0.5)\n",
    "\n",
    "temp = np.log(list_CIS)\n",
    "scorr = spearmanr(list_CAI, temp)\n",
    "pcorr = pearsonr(list_CAI, temp)\n",
    "print(pcorr)\n",
    "print(scorr)\n",
    "ax1.scatter(list_CAI, list_CIS, color=\"C4\", marker=\".\", s=200)\n",
    "ax1.set_xlabel(used_info, fontsize=20)\n",
    "ax1.set_ylabel(\"CIS$_{c,\\cdot}$\", fontsize=20)\n",
    "ax1.set_title(\"CIS vs. {}\\nin {}\".format(used_info, RiboMIMONet.dataset), fontsize=20, pad=20)\n",
    "ax1.set_yscale(\"log\")\n",
    "ax1.spines[\"top\"].set_visible(False)\n",
    "ax1.spines[\"right\"].set_visible(False)\n",
    "ax1.spines[\"left\"].set_position(('outward', 5))\n",
    "ax1.spines[\"bottom\"].set_position(('outward', 5))\n",
    "ax1.spines[\"bottom\"].set_bounds(0, 1)\n",
    "ax1.spines[\"left\"].set_bounds(1e-3, 1e-1)\n",
    "ax1.axis([-0.1, 1.1, 1e-3, 1e-1])\n",
    "for tick in ax1.xaxis.get_major_ticks():\n",
    "    tick.label.set_fontsize(14) \n",
    "for tick in ax1.yaxis.get_major_ticks():\n",
    "    tick.label.set_fontsize(14) \n",
    "axis = ax1.axis()\n",
    "ax1.text(axis[0]+0.3*(axis[1]-axis[0]), \n",
    "         axis[3]/np.exp(0.05*np.log(axis[3]/axis[2])), \n",
    "         \"scc: ${:.4f}$ (p-value: ${:.1g}$)\".format(scorr[0], scorr[1]), \n",
    "         fontsize=14, horizontalalignment='left', verticalalignment='center')\n",
    "ax1.text(axis[0]+0.3*(axis[1]-axis[0]), \n",
    "         axis[3]/np.exp(0.1*np.log(axis[3]/axis[2])), \n",
    "         \"pcc: ${:.4f}$ (p-value: ${:.1g}$)\".format(pcorr[0], pcorr[1]), \n",
    "         fontsize=14, horizontalalignment='left', verticalalignment='center')\n",
    "ax1.text(axis[0]-0.2*(axis[1]-axis[0]), \n",
    "         axis[3]*np.exp(0.2*np.log(axis[3]/axis[2])),\n",
    "         \"D\", fontsize=20, horizontalalignment='center', verticalalignment='center')\n",
    "ax1.yaxis.set_label_coords(-0.2, 0.5)\n",
    "\n",
    "plt.subplots_adjust(left=0.125, bottom=0.1, right=0.9, top=0.9, wspace=0.3, hspace=0.2)\n",
    "# plt.show()\n",
    "plt.savefig(\"{}/Fig3_CAI_{}.pdf\".format(RiboMIMONet.save_path, RiboMIMONet.dataset), bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 RPF regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 10 fold RPF saliency map\n",
    "window = 10\n",
    "dict_influence_codon = np.zeros((61, window))\n",
    "counts = np.zeros((61, window))\n",
    "for idx in range(len(RiboMIMONet.data.list_gene)):\n",
    "    print(\"{:.2f}%\\r\".format((1.0+idx)/len(RiboMIMONet.data.list_gene)*100), end=\"\")\n",
    "    gene_name = RiboMIMONet.data.list_gene[idx]\n",
    "    seq = RiboMIMONet.data.dict_seq[gene_name]\n",
    "    len_codon = len(seq)//3\n",
    "    codons = [seq[i*3:(i+1)*3] for i in range(len_codon)]\n",
    "    try:\n",
    "        saliency_sum = pickle.load(open(\"{}/{}\".format(cis_path, gene_name), \"rb\"))\n",
    "    except:\n",
    "        continue\n",
    "    sm_2d = saliency_sum\n",
    "    for i in range(len_codon):\n",
    "        temp_codon = codons[i]\n",
    "        if temp_codon in [\"TAA\", \"TAG\", \"TGA\"]:\n",
    "            continue\n",
    "        start = max(0, i-window//2+1)\n",
    "        stop = min(len_codon, i+window//2+1)\n",
    "        dict_influence_codon[codon2idx[temp_codon], window//2+1+i-stop:window//2+1+i-start] += sm_2d[start:stop, i][::-1]\n",
    "        counts[codon2idx[temp_codon], window//2+1+i-stop:window//2+1+i-start] += 1\n",
    "dict_influence_codon /= counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.rcParams['font.family'] = 'sans-serif'\n",
    "y, x = np.meshgrid(np.array(range(window+1)), np.array(range(62)))\n",
    "# x and y are bounds, so z should be the value *inside* those bounds.\n",
    "# Therefore, remove the last value from the z array.\n",
    "# z = np.log(dict_influence_center.copy()+1)\n",
    "z = dict_influence_codon.copy()\n",
    "z_min, z_max = -np.max(np.abs(z)), np.max(np.abs(z))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(15, 3), dpi=200)\n",
    "c = ax.pcolormesh(x, y, z, cmap='RdBu_r', vmin=z_min, vmax=z_max)\n",
    "fig.colorbar(c, ax=ax)\n",
    "ax.set_ylabel(\"Offset from A site (codon)\", fontsize=12)\n",
    "ax.set_title(\"CIS at different codon positions \" + r\"in {}\".format(RiboMIMONet.dataset))\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "ax.spines[\"bottom\"].set_visible(False)\n",
    "ax.spines[\"left\"].set_visible(False)\n",
    "ax.tick_params(axis=\"x\", labelcolor=\"w\", colors=\"w\", rotation=90)\n",
    "ax.tick_params(axis=\"y\", labelcolor=\"k\", colors=\"w\", rotation=90)\n",
    "\n",
    "list_yticks = np.array([str(i) for i in range(-window//2, window//2)])\n",
    "list_yticks[window//2-2:window//2+1] = [\"E\", \"P\", \"A\"]\n",
    "ax.set_yticks(np.array(range(window))+0.5)\n",
    "ax.set_yticklabels(list_yticks)\n",
    "ax.text(-3, 11, \"E\", fontsize=12, horizontalalignment='center', verticalalignment='center')\n",
    "\n",
    "for i in range(len(list_codon)):\n",
    "    ax.text(i+0.5, -2, list_codon[i], rotation=90, fontsize=10, horizontalalignment='center', verticalalignment='center')\n",
    "    \n",
    "# plot aa\n",
    "\n",
    "aa_plot = []\n",
    "current = None\n",
    "count = 0\n",
    "for i in range(len(list_aa)):\n",
    "    if list_aa[i] == current:\n",
    "        count += 1\n",
    "    else:\n",
    "        aa_plot.append([current, i-count, count])\n",
    "        current = list_aa[i]\n",
    "        count = 1\n",
    "aa_plot.append([current, i-count+1, count])\n",
    "for i in aa_plot:\n",
    "    if i[2] == 0:\n",
    "        continue\n",
    "    ax.plot([i[1]+0.3, i[1]+i[2]-0.3], [-3.5, -3.5], color=\"k\", lw=3)\n",
    "    ax.text(i[1]+i[2]*0.5, -4.5, i[0], fontsize=12, rotation=90, horizontalalignment='center', verticalalignment='center')\n",
    "\n",
    "ax.yaxis.set_label_coords(-0.03, 0.5)\n",
    "# plt.show()\n",
    "plt.savefig(\"{}/Fig4_RPF_codon_{}.pdf\".format(RiboMIMONet.save_path, RiboMIMONet.dataset), bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 10 fold RPF saliency map\n",
    "window = 10\n",
    "dict_influence_aa = np.zeros((20, window))\n",
    "counts = np.zeros((20, window))\n",
    "for idx in range(len(RiboMIMONet.data.list_gene)):\n",
    "    print(\"{:.2f}%\\r\".format((1.0+idx)/len(RiboMIMONet.data.list_gene)*100), end=\"\")\n",
    "    gene_name = RiboMIMONet.data.list_gene[idx]\n",
    "    seq = RiboMIMONet.data.dict_seq[gene_name]\n",
    "    len_codon = len(seq)//3\n",
    "    codons = [seq[i*3:(i+1)*3] for i in range(len_codon)]\n",
    "    try:\n",
    "        saliency_sum = pickle.load(open(\"{}/{}\".format(cis_path, gene_name), \"rb\"))\n",
    "    except:\n",
    "        continue\n",
    "    sm_2d = saliency_sum\n",
    "    for i in range(len_codon):\n",
    "        temp_codon = codons[i]\n",
    "        if temp_codon in [\"TAA\", \"TAG\", \"TGA\"]:\n",
    "            continue\n",
    "        temp_aa = RiboMIMONet.data.codon2aa[temp_codon]\n",
    "        start = max(0, i-window//2+1)\n",
    "        stop = min(len_codon, i+window//2+1)\n",
    "        dict_influence_aa[aa2idx[temp_aa], window//2+1+i-stop:window//2+1+i-start] += sm_2d[start:stop, i][::-1]\n",
    "        counts[aa2idx[temp_aa], window//2+1+i-stop:window//2+1+i-start] += 1\n",
    "dict_influence_aa /= counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.rcParams['font.family'] = 'sans-serif'\n",
    "y, x = np.meshgrid(np.array(range(window+1)), np.array(range(21)))\n",
    "# x and y are bounds, so z should be the value *inside* those bounds.\n",
    "# Therefore, remove the last value from the z array.\n",
    "# z = np.log(dict_influence_center.copy()+1)\n",
    "z = dict_influence_aa.copy()\n",
    "z_min, z_max = -np.max(np.abs(z)), np.max(np.abs(z))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 2), dpi=200)\n",
    "c = ax.pcolormesh(x, y, z, cmap='RdBu_r', vmin=z_min, vmax=z_max)\n",
    "fig.colorbar(c, ax=ax)\n",
    "ax.set_ylabel(\"Offset from A site\\n(codon)\", fontsize=10)\n",
    "ax.set_title(\"CIS at different amino acid positions\\n\" + r\"in {}\".format(RiboMIMONet.dataset),\n",
    "             fontsize=10)\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "ax.spines[\"bottom\"].set_visible(False)\n",
    "ax.spines[\"left\"].set_visible(False)\n",
    "ax.tick_params(axis=\"x\", labelcolor=\"w\", colors=\"w\", rotation=90)\n",
    "ax.tick_params(axis=\"y\", labelcolor=\"k\", colors=\"w\", rotation=90)\n",
    "\n",
    "list_yticks = np.array([str(i) for i in range(-window//2, window//2)])\n",
    "list_yticks[window//2-2:window//2+1] = [\"E\", \"P\", \"A\"]\n",
    "ax.set_yticks(np.array(range(window))+0.5)\n",
    "ax.set_yticklabels(list_yticks)\n",
    "ax.text(-4, 13, \"F\", fontsize=10, horizontalalignment='center', verticalalignment='center')\n",
    "\n",
    "for i in range(len(list_aa_unique)):\n",
    "    ax.text(i+0.5, -0.5, list_aa_unique[i], rotation=90, fontsize=8, horizontalalignment='center', verticalalignment='center')\n",
    "\n",
    "ax.yaxis.set_label_coords(-0.1, 0.5)\n",
    "# plt.show()\n",
    "plt.savefig(\"{}/Fig4_RPF_aa_{}.pdf\".format(RiboMIMONet.save_path, RiboMIMONet.dataset), bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.4 Long-range contributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 10 fold pair saliency map\n",
    "dict_influence_pair = np.zeros((61, 61))\n",
    "counts = np.zeros((61, 61))\n",
    "window = 20\n",
    "for idx in range(len(RiboMIMONet.data.list_gene)):\n",
    "    print(\"{:.2f}%\\r\".format((1.0+idx)/len(RiboMIMONet.data.list_gene)*100), end=\"\")\n",
    "    gene_name = RiboMIMONet.data.list_gene[idx]\n",
    "    seq = RiboMIMONet.data.dict_seq[gene_name]\n",
    "    len_codon = len(seq)//3\n",
    "    codons = [seq[i*3:(i+1)*3] for i in range(len_codon)]\n",
    "    try:\n",
    "        saliency_sum = pickle.load(open(\"{}/{}\".format(cis_path, gene_name), \"rb\"))\n",
    "    except:\n",
    "        continue\n",
    "    sm_2d = saliency_sum\n",
    "    for i, j in np.array(np.triu_indices(len_codon-1, window)).T:\n",
    "        a = codons[i]\n",
    "        if a in [\"TAA\", \"TAG\", \"TGA\"]:\n",
    "            continue\n",
    "        b = codons[j]\n",
    "        if b in [\"TAA\", \"TAG\", \"TGA\"]:\n",
    "            continue\n",
    "        idxa = codon2idx[a]\n",
    "        idxb = codon2idx[b]\n",
    "        # left\n",
    "        dict_influence_pair[idxa, idxb] += sm_2d[j, i]\n",
    "        counts[idxa, idxb] += 1\n",
    "        # right\n",
    "        dict_influence_pair[idxb, idxa] += sm_2d[i, j]\n",
    "        counts[idxb, idxa] += 1\n",
    "dict_influence_pair /= counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "z = dict_influence_pair\n",
    "y, x = np.meshgrid(np.array(range(z.shape[0]+1)), np.array(range(z.shape[1]+1)))\n",
    "# x and y are bounds, so z should be the value *inside* those bounds.\n",
    "# Therefore, remove the last value from the z array.\n",
    "z_min, z_max = -np.max(np.abs(z)), np.max(np.abs(z))\n",
    "# z_min, z_max = np.min(z), np.max(z)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(15, 12), dpi=200)\n",
    "c = ax.pcolormesh(x, y, z[::-1], cmap='RdBu_r', vmin=z_min, vmax=z_max)\n",
    "ax.axis([x.min(), x.max(), y.min(), y.max()])\n",
    "fig.colorbar(c, ax=ax)\n",
    "ax.set_title(\"Impact scores for long-range codon pairs\\nin {}\".format(RiboMIMONet.dataset), fontsize=25, pad=20)\n",
    "ax.set_xlabel(\"Donor\", fontsize=20)\n",
    "ax.set_ylabel(\"Acceptor\", fontsize=20)\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "ax.spines[\"bottom\"].set_visible(False)\n",
    "ax.spines[\"left\"].set_visible(False)\n",
    "ax.tick_params(axis=\"x\", labelcolor=\"k\", colors=\"w\")\n",
    "ax.tick_params(axis=\"y\", labelcolor=\"k\", colors=\"w\")\n",
    "plt.xticks(np.array(range(len(z)))+0.5, list_codon[::-1], rotation=90)\n",
    "plt.yticks(np.array(range(len(z)))+0.5, list_codon, rotation=0)\n",
    "#plt.rcParams['font.family'] = 'monospace'\n",
    "for tick in ax.get_xticklabels():\n",
    "    tick.set_fontname(\"Consolas\")\n",
    "for tick in ax.get_yticklabels():\n",
    "    tick.set_fontname(\"Consolas\")\n",
    "plt.text(-2, 64, \"G\", fontsize=20, horizontalalignment='center', verticalalignment='center')\n",
    "# plt.show()\n",
    "plt.savefig(\"{}/Fig5_long_range_{}.pdf\".format(RiboMIMONet.save_path, RiboMIMONet.dataset), bbox_inches = 'tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
